File: leetcode/2024_8_august/c++/16_maximum_distance_in_arrays.cpp
// TIME COMPEXITY
// O(n)

// SPACE COMPEXITY
// O(1)

// CODE

class Solution {
public:
    int maxDistance(vector<vector<int>>& arrays) {
        int miniboi = arrays[0].front();
        int maxiboi = arrays[0].back();

        int res = 0;

        for(int i = 1; i < arrays.size(); i++){
            int currMin = arrays[i].front();
            int currMax = arrays[i].back();

            res = max({res, abs(currMin - maxiboi), abs(currMax - miniboi)});

            miniboi = min(miniboi, currMin);
            maxiboi = max(maxiboi, currMax);
        }
        return res;
    }
};

File: leetcode/2024_8_august/python/16_maximum_distance_in_arrays.py
# CODE

from typing import List

class Solution:
    def maxDistance(self, arrays: List[List[int]]) -> int:
        res = 0
        maxiboi = arrays[0][-1]
        miniboi = arrays[0][0]
        for i in range(1, len(arrays)):
            first = arrays[i][0]
            last = arrays[i][-1]
            res = max(last - miniboi, maxiboi - first, res)
            maxiboi = max(maxiboi, last)
            miniboi = min(miniboi, first)
        return res